### ISSUE TITLE
Model object in the field

### ISSUE THREAD

**Issue by Tensho** [thoughtbot/administrate#267]
I've found that resource (record) disappears, while administrate converts show page attributes to the fields.

```
# administrate-0.1.1/lib/administrate/page/base.rb
…

protected

def attribute_field(dashboard, resource, attribute_name, page)
  value = get_attribute_value(resource, attribute_name)

  dashboard.
    attribute_types[attribute_name].
    new(attribute_name, value, page)
end
```

In some cases it’s really necessary to take reference to the object, some helpers require model. For example, Administrate::Page::Show instance looses such public reference during the rendering. That’s why I have to use some hack, when I want to provide Refile attachment url (which requires record object) during the template rendering:

```
# app/views/fields/file_field/_show.html.erb
<% record = assigns['_requested_resource'] %>
<%= image_tag attachment_url(record, field.attribute, :fill, 300, 300, format: "jpg") %>
```

What do you think about the record exposing, making it public via field object? Does it violate SRP (or any other good principles) of the field class design?

**Comment by graysonwright** [comment]
@Tensho - let me make sure I understand:

You're creating a custom field type called `FileField`, and the `show.html.erb` partial for the field needs a reference to the resource itself, and not just the attribute value. Is that right?

If that's the case, there's a workaround:

First, run `rails generate administrate:views:show`. This will generate a file at `app/views/admin/application/show.html.erb`.

On line 35 of the generated file, change the line that looks like:

```eruby
<%= render_field attribute %>
```

to:

```eruby
<%= render_field attribute, resource: page.resource %>
```

Now, inside of `app/views/fields/file_field/_show.html.erb`, you'll have access to the `resource` variable, which is the ActiveRecord object.

Let me know if that works for you, or if I made a mistake somewhere. Cheers!

**Comment by Tensho** [comment]
Yeah, you're right. Actually, I've already found this solution by myself for `index` partial:
```
# /app/helpers/administrate/application_helper.rb
module Administrate
  module ApplicationHelper
    def render_field(field, resource = nil, locals = {})
      locals.merge!(resource: resource, field: field)
      render locals: locals, partial: field.to_partial_path
    end
  end
end

# /app/views/administrate/application/_collection.html.erb
<td class="cell-data cell-data--<%= attribute.html_class %>">
  <%= render_field attribute, resource %>
</td>
```
My main question is related to design concerns. Are you going to hide resource/record/model further and substitute it with the field object during the rendering process intentionally?
I can provide PR with the resource bubbling to the view layer, something similar we're talking about here. However, it seems to me you need to consider resource object reference in the field object. I know, there are not so many (non-form) helpers require record, but there is a case. Probably other guys will encounter with this problem within other gems.
If you don't see any necessity to include the info about record inside rendering process (particularly, field), just close this issue.

**Comment by Tensho** [comment]
I also can take care about my proposal in the form of PR ^_^ if you accept it and advice the best approach.

**Comment by graysonwright** [comment]
@Tensho that looks great!

I'm open to making this change, but I want to gauge interest first. If other people run into this problem as well, we can definitely add it in. I'll leave this issue open for now so people can comment on it.

**Comment by AlexBeznos** [comment]
:+1:

**Comment by mcmire** [comment]
I'm running into a case where I have a field and I would like to have access to the resource object as well. It's sort of a similar case here -- my model stores a reference to an audio file and I'd like to render a player for that file. The attribute in question is Carrierwave-enabled, so the only way to get access to the URL is by saying `resource.[attribute].url`.

**Comment by dts** [comment]
In principle, I'd like access to the resource from the field object as well as from the view.  In my PR (#706) I add support for this.

**Comment by avk** [comment]
I just ran into this case in trying to display a custom field in a format that relies on model logic. Would love to see this in administrate 👍

**Comment by kevinelliott** [comment]
I'd like to see this implemented.

**Comment by icstunna** [comment]
I too would like to see this implemented. Preferably with multiple resources.

**Comment by dts** [comment]
@avk, @kevine, @icstunna: see my PR (#706) and post about whether it is would solve your use case, and/or if the API is sensible.

**Comment by carlosramireziii** [comment]
@dts [Your approach](https://github.com/thoughtbot/administrate/pull/706/files) seems to be a bit different than [@Tensho's](https://github.com/Tensho/administrate/commit/2480b9f4d3fc83cf8678cfe74c18763eada84b04). Could you describe the differences? Does your PR incorporate other changes as well?

**Comment by dts** [comment]
@carlosramireziii - It looks like @Tensho's solution does not pass the resource into the actual field object, only in the view side.  I believe it's important to have access to the resource in question from the field object, so it's available in all the places you might want to access it.

**Comment by nickcharlton** [comment]
I _think_ this issue was solved with #706, so I'm going to close this. But please open a new one if that's not the case.
